### ISSUE TITLE
How to improve recall rate on specific class?

### ISSUE THREAD

**Issue by zero90169** [ultralytics/yolov5#1995]
## ❔Question

First, thanks to the author for sharing this amazing repo.

I have used yolov5s and trained on my custom dataset. Due to inference speed reason so I set the image size to 320x320 ( I know it is too small... ) The overall precision (70%) looks fine, almost the same (or even higher) with the another model MobileDet. However the overall recall performs bad (compare with the MobileDet which got 70% recall, yolov5s only got 60%).

I have tried to use focal loss, but the training process become very long. I want to know is there has another trick to enhance the recall rate or I can improve the recall rate on specific class via some methods (different class with different weighted)?

**Comment by ZhWL123456** [comment]
Maybe data augment for the specific class is helpful.

**Comment by glenn-jocher** [comment]
@zero90169 mAP is by far the best metric to track, as it provides _absolute_ performance across all confidence levels. Precision and recall are _relative_ metrics that vary as a function of confidence threshold. Therefore do not concern yourself with P and R, focus on mAP for an apples to apples comparison.

Best results are achieved when training and detecting at the same --img size, so if you want to deploy with 320 inference size, you should train at 320 as well.

Individual class weights can be specified in the class_weights vector, and weighted classification loss can then employ the class weights in the ComputeLoss() class. Note that class weights are defined as inverse frequencies by default.
https://github.com/ultralytics/yolov5/blob/d9212140b355b84e85a473be590720eb8221766c/train.py#L219

**Comment by zero90169** [comment]
@glenn-jocher Thanks for your suggestions. I will try the class_weights vector.

I did train on 320x320 and inference my well-trained model on 320x320. I knew that R and P will be affected by the threshold, but in my application scenario I have to focus on R in some class (given threshold=0.5).

p.s. I found that in detect.py the input image will be resized to 192x320 not 320x320, it is different to the image we used in training process 320x320. I don't know whether it will affect to the performance or not.

**Comment by glenn-jocher** [comment]
@zero90169 if you really need to focus on R, make sure you set a reasonable threshold then for your data here (default is 0.1, this value is different than than --conf)
https://github.com/ultralytics/yolov5/blob/046c37e465d60b9047ff0e91f44594817e5a935c/utils/metrics.py#L41

**Comment by ggyybb** [comment]
> @zero90169 mAP is by far the best metric to track, as it provides _absolute_ performance across all confidence levels. Precision and recall are _relative_ metrics that vary as a function of confidence threshold. Therefore do not concern yourself with P and R, focus on mAP for an apples to apples comparison.
> 
> Best results are achieved when training and detecting at the same --img size, so if you want to deploy with 320 inference size, you should train at 320 as well.
> 
> Individual class weights can be specified in the class_weights vector, and weighted classification loss can then employ the class weights in the ComputeLoss() class. Note that class weights are defined as inverse frequencies by default.
> https://github.com/ultralytics/yolov5/blob/d9212140b355b84e85a473be590720eb8221766c/train.py#L219

I am interested in model.class_weights, I konw how to compute this weights vector, but in loss.py ,I can't find where to use this model.class_weights, please help me understand this place at your convenience,thanks a lot!

**Comment by glenn-jocher** [comment]
@ggyybb model.class_weights can be applied in the classification BLE loss function here:
https://github.com/ultralytics/yolov5/blob/d2e754b67bc08d3634df05932cc94d8c9314a7b1/utils/loss.py#L96

The weights can be passed to the loss function to define a value for the `pos_weights` argument. See documentation for details:
https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html

**Comment by ggyybb** [comment]
> @ggyybb model.class_weights can be applied in the classification BLE loss function here:
> https://github.com/ultralytics/yolov5/blob/d2e754b67bc08d3634df05932cc94d8c9314a7b1/utils/loss.py#L96
> 
> The weights can be passed to the loss function to define a value for the `pos_weights` argument. See documentation for details:
> https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html

![image](https://user-images.githubusercontent.com/61568153/108443609-12bc7900-7294-11eb-8b59-a7f7c82a3cfc.png)
I think the cls_pw in hyp do not work actually, it's just a number, if I modified cls_pw in hyp.yaml, the pos_weights of BCEcls will change, but will not work actually, is my idea right?
So I use model.class_weights in pos_weights, It work, I hope i am right, am I? Thanks so much for your response!

**Comment by glenn-jocher** [comment]
@ggyybb that's a good question. The interpretation of the pos_weight argument seems up for debate based on a few forum responses I saw. We were treating it as 'positive weight' before that I assumed was applied to positive samples, but the current documentation shows this as 'position weight'.

You may want to experiment with this to see what the true effect is.

**Comment by glenn-jocher** [comment]
@ggyybb BTW I'm testing `pos_weight=model.class_weights` on YOLOv5m COCO, and in early training so far it's trailing the default result, but I'll need a few more days to have a better picture. 

To be clear, the `model.class_weights` vector is definitely created correctly, I spent a lot of time getting this right. It is an inverse frequency vector that sums to the class count (80). The main question is whether it can be applied directly using `pos_weight=model.class_weights`, or whether you may need to apply it manually by setting reduction differently and multiplying the loss matrix by this row vector before doing the reduction. You'll need to experiment.

**Comment by ggyybb** [comment]
> @ggyybb BTW I'm testing `pos_weight=model.class_weights` on YOLOv5m COCO, and in early training so far it's trailing the default result, but I'll need a few more days to have a better picture.
> 
> To be clear, the `model.class_weights` vector is definitely created correctly, I spent a lot of time getting this right. It is an inverse frequency vector that sums to the class count (80). The main question is whether it can be applied directly using `pos_weight=model.class_weights`, or whether you may need to apply it manually by setting reduction differently and multiplying the loss matrix by this row vector before doing the reduction. You'll need to experiment.

I studied the source code carefully,I think the pos_weights para is positive weights, It's same as you thought before, It should be a number ,or a hyp, and it will broadcast to fit loss function.
And another para 'weight' of nn.BCEWithLogistLoss is the weights of each class, the doc of torch say:'weight (Tensor, optional): a manual rescaling weight given to the loss of each batch element'.so, I think the the 'each batch element' means that 'class',and I do some experiment with some random number, weight=model.class_weights will apply to each class when the loss was computed.
So, I think make weight=model.class_weights maybe the right use of model.class_weights.

**Comment by ggyybb** [comment]
> @ggyybb BTW I'm testing `pos_weight=model.class_weights` on YOLOv5m COCO, and in early training so far it's trailing the default result, but I'll need a few more days to have a better picture.
> 
> To be clear, the `model.class_weights` vector is definitely created correctly, I spent a lot of time getting this right. It is an inverse frequency vector that sums to the class count (80). The main question is whether it can be applied directly using `pos_weight=model.class_weights`, or whether you may need to apply it manually by setting reduction differently and multiplying the loss matrix by this row vector before doing the reduction. You'll need to experiment.

Like this
![image](https://user-images.githubusercontent.com/61568153/108487280-0d821d00-72da-11eb-8042-80cc7953b1f6.png)

**Comment by glenn-jocher** [comment]
@ggyybb oh good work. That's what I thought originally. So I will cancel my current training, which is doing `pos_weight=model.class_weights` and start a new one with `weight=model.class_weights.`

**Comment by glenn-jocher** [comment]
@ggyybb just to check, the behavior with `weight=model.class_weights` you are observing is that each loss column (1 class per column) is being multiplied by it's model.class_weights value right? 

An example loss shape for COCO might be 1000x80, where all of the 1000 values in the first column should be multiplied by model.class_weights[0].

**Comment by ggyybb** [comment]
> @ggyybb just to check, the behavior with `weight=model.class_weights` you are observing is that each loss column (1 class per column) is being multiplied by it's model.class_weights value right?
> 
> An example loss shape for COCO might be 1000x80, where all of the 1000 values in the first column should be multiplied by model.class_weights[0].

I think Its right,I will be looking forward to your new training result, and I will do experiment on my own datasets,Thanks very much!

**Comment by glenn-jocher** [comment]
@ggyybb looking like this so far. cls loss is down by 50% because of the weightings, and mAP is a bit lower too in early training. Much of the current effect may simply due to a lower cls loss in general now though rather than the actual weightings, i.e. it may make sense to increase hyp['cls']=1.5 or 2.0 to rebalance the losses after the change. In any case too early to tell, maybe 3 days left of training.

<img width="1014" alt="Screen Shot 2021-02-21 at 11 13 24 AM" src="https://user-images.githubusercontent.com/26833433/108635735-d4f75600-7435-11eb-88a4-6f650f0cf461.png">

**Comment by glenn-jocher** [comment]
@ggyybb comparison is done. class_weights (red) reduces mAP about -.01 for COCO YOLOv5m vs default (green) in my experiment.

<img width="1000" alt="Screen Shot 2021-02-24 at 12 15 06 AM" src="https://user-images.githubusercontent.com/26833433/108970860-e8254400-7636-11eb-90e1-ebef1f1cd57f.png">

**Comment by ShirleyHe2020** [comment]
> > @ggyybb BTW I'm testing `pos_weight=model.class_weights` on YOLOv5m COCO, and in early training so far it's trailing the default result, but I'll need a few more days to have a better picture.
> > To be clear, the `model.class_weights` vector is definitely created correctly, I spent a lot of time getting this right. It is an inverse frequency vector that sums to the class count (80). The main question is whether it can be applied directly using `pos_weight=model.class_weights`, or whether you may need to apply it manually by setting reduction differently and multiplying the loss matrix by this row vector before doing the reduction. You'll need to experiment.
> 
> Like this
> ![image](https://user-images.githubusercontent.com/61568153/108487280-0d821d00-72da-11eb-8042-80cc7953b1f6.png)

Hi 

> > @ggyybb BTW I'm testing `pos_weight=model.class_weights` on YOLOv5m COCO, and in early training so far it's trailing the default result, but I'll need a few more days to have a better picture.
> > To be clear, the `model.class_weights` vector is definitely created correctly, I spent a lot of time getting this right. It is an inverse frequency vector that sums to the class count (80). The main question is whether it can be applied directly using `pos_weight=model.class_weights`, or whether you may need to apply it manually by setting reduction differently and multiplying the loss matrix by this row vector before doing the reduction. You'll need to experiment.
> 
> Like this
> ![image](https://user-images.githubusercontent.com/61568153/108487280-0d821d00-72da-11eb-8042-80cc7953b1f6.png)

Hi, can you please share your loss_read.py  as shown in the pic ?
